docs/P4c-test-documentation.md

```diff
@@ -0,0 +1,235 @@
+# **Adding New Tests to the P4C Repository**
+
+**This document provides a guide for developers who want to add new tests to the P4C repository. It explains where to place test files, how to manage expected output files, and how to handle tests that are expected to fail (Xfail).**
+
+---
+
+
+
+
+
+## **1. Adding "Positive" Tests**
+
+Positive tests are those where the P4 compiler should process the input without any errors.
+
+### Steps for Adding Positive Tests
+
+1. **Source File Location**:
+
+   - Positive tests are located in multiple directories based on the purpose and the backend/tool being tested. Below is a description of the primary directories under `p4c/testdata` and their purposes:
+
+     - **`p4_16_samples`**: Contains general positive tests for the P4\_16 language.
+     - **`p4_16_samples_outputs`**: Stores the expected outputs for the tests in `p4_16_samples`.
+     - **`p4_14_samples`**: Contains tests specific to the older P4\_14 language version.
+     - **`p4_14_samples_outputs`**: Stores the expected outputs for tests in `p4_14_samples`.
+     - **`bmv2`**: Holds tests targeting the BMv2 backend, including packet-processing logic and examples.
+     - **`psa`**: Includes tests for the PSA (Portable Switch Architecture) target.
+     - **`ebpf`**: Contains tests designed for the eBPF backend.
+     - **`xdp`**: Stores tests related to XDP programs using eBPF.
+     - **`arch`**: Contains architecture-specific tests, covering various P4 architectures.
+     - **`tools`**: Includes tests for tools provided in the P4C repository, such as `p4test` and `p4c-bm2-ss`.
+     - **`error_samples`**: Contains tests that verify the compiler's behavior when encountering erroneous P4 programs.
+     - **`stf_samples`**: Stores `.stf` (Switch Test Framework) files associated with P4 programs, defining packet inputs/outputs for testing.
+     - **`stress_tests`**: Contains larger or more computationally intensive tests to stress-test the compiler.
+
+   - Make sure the file you are adding is placed in the appropriate directory based on the purpose and backend/tool it is testing.
+   - Ensure the file is named descriptively to reflect its purpose (e.g., `example_positive_test.p4`).
+
+
+2. **File Naming Conventions**:
+
+   - For **BMv2-specific tests** with packets, the source file name must match the pattern `*-bmv2.p4`.
+  - For **v1model architecture** tests: `*-bmv2-v1model.p4`.
```

Comment:
There are currently _no_ test files with a file name that ends with `-bmv2-v1model.p4`.  If you tried to create one, it would not match the pattern `*-bmv2.p4`, and thus the cmakefiles and existing test infrastructure would _not_ run BMv2 on such a test program.